source/src/BNFC/Backend/Java/CFtoJLex15.hs

```diff
@@ -60,28 +60,55 @@ cf2jlex jflex packageBase cf = (vcat
    makeSymEnv (s:symbs) n = (s, "_SYMB_" ++ show n) : makeSymEnv symbs (n+1)
 
 -- | File prelude
-prelude :: Bool -> String -> Doc
-prelude jflex packageBase = vcat
+prelude :: (Bool, Bool) -> String -> Doc
+prelude (ln, jflex) packageBase = vcat
     [ "// This JLex file was machine-generated by the BNF converter"
     , "package" <+> text packageBase <> ";"
     , ""
     , "import java_cup.runtime.*;"
     , "%%"
     , "%cup"
     , "%unicode"
-    , "%line"
+    , (if ln 
+      then vcat
+        [ "%line"
+        , (if jflex then "%column" else "")
+        , "%char" ]
+      else "")
     , "%public"
     , "%{"
     , nest 2 $ vcat
         [ "String pstring = new String();"
+        , "ComplexSymbolFactory.Location left = new ComplexSymbolFactory.Location(-1, -1);"
+        , "ComplexSymbolFactory cf = new ComplexSymbolFactory();"
+        , positionDeclarations
         , "public int line_num() { return (yyline+1); }"
+        , "public ComplexSymbolFactory.Location left_loc() {"
+        , "  return new ComplexSymbolFactory.Location(yyline+1, yycolumn+1, yychar+1);"
+        , "}"
+        , "public ComplexSymbolFactory.Location right_loc() {"
+        , "  ComplexSymbolFactory.Location right = left_loc();"
+        , (if ln
+            then "right.move(0, yylength(), yylength());"
+            else "")
+        , "  return right;"
+        , "}"
         , "public String buff()" <+> braces
             (if jflex
             then "return new String(zzBuffer,zzCurrentPos,10).trim();"
             else "return new String(yy_buffer,yy_buffer_index,10).trim();")
         ]
     , "%}"
     ]
+  where
+    positionDeclarations =
+      if ln
+        then if jflex then ""
+                      else "int yycolumn = -2;"
+        else vcat
+          [ "int yyline = -2;"
```

Comment:
Magic numbers like `-2` should be avoided.
If unavoidable, document properly. (Why `-2`, why doesn't it work without the magic?)

Reply:
Done.